Mono: Add Dictionary and Array classes

neikeq opened the pull request:
Please test :)

Any opinion on the way I did Array <---> Array\<T\>? (and Dictionary)

neikeq commented on core/variant_call.cpp:
```diff
@@ -1606,7 +1606,7 @@ void register_variant_methods() {
 	ADDFUNC1R(DICTIONARY, BOOL, Dictionary, has, NIL, "key", varray());
 	ADDFUNC1R(DICTIONARY, BOOL, Dictionary, has_all, ARRAY, "keys", varray());
-	ADDFUNC1(DICTIONARY, NIL, Dictionary, erase, NIL, "key", varray());
+	ADDFUNC1R(DICTIONARY, BOOL, Dictionary, erase, NIL, "key", varray());
```
This breaks API compatibility. If this is unacceptable at this point I can replace my usage with `if has_key then erase` but that would be a bit slower :)

neikeq commented:
I better fix these conflicts before Akien starts discriminating against my branch :fearful:

reduz commented:
To me seems ok

On Jan 19, 2018 7:46 PM, "Ignacio Etcheverry" <notifications@github.com>
wrote:

> I better fix these conflicts before Akien starts discriminating against my
> branch 😨
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/15880#issuecomment-359110941>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z21feoT0pvHHVKH2FMwhdD-3o1fp0ks5tMRtLgaJpZM4RlJlG>
> .
>

cart commented:
This should give a nice performance boost :)  Looks good to me.

akien-mga commented:
> I better fix these conflicts before Akien starts discriminating against my branch 😨

Well you better fix those conflits before 3.0-stable mostly... And ideally before making a PR, a rebase is not hard ;)

karroffel commented:
Can you modify the GDNative bindings for Dictionary too?

https://github.com/godotengine/godot/blob/09732a15a50cedc4d495588f77784c26d3b3c7a7/modules/gdnative/include/gdnative/dictionary.h#L75-L77

https://github.com/godotengine/godot/blob/09732a15a50cedc4d495588f77784c26d3b3c7a7/modules/gdnative/gdnative/dictionary.cpp#L85-L89

https://github.com/godotengine/godot/blob/09732a15a50cedc4d495588f77784c26d3b3c7a7/modules/gdnative/gdnative_api.json#L2866-L2873

That should be it.

ReDucTor commented on modules/mono/glue/collections_glue.cpp:
```diff
@@ -0,0 +1,212 @@
+	Variant varKey = GDMonoMarshal::mono_object_to_variant(key);
+	Variant *ret = ptr->getptr(varKey);
+	if (ret != NULL)
+		mono_raise_exception(mono_get_exception_argument("key", "An element with the same key already exists"));
```
I haven't looked at godot general implementation of mono embedding, but you need to be careful using mono_raise_exception as it is not going to finish the current stack frame, so in this case it would cause 'varKey' destructor not to be called

neikeq replied:
Thanks. Indeed, that would have messed things up. Now it's using pending exceptions instead.

akien-mga commented:
@neikeq So when do we get those conflicts fixed? ;)

akien-mga commented:
So, what's up with this PR?

neikeq commented:
Not the first in my priority list right now. I need to add PoolArrays as well and decide what to do with exceptions.

aaronfranke commented:
What are the advantages gained in this PR over C#'s built-in Dictionaries and Arrays?

neikeq commented:
@aaronfranke Array and Dictionary are mutable in Godot. So are C#'s Dictionary and List, but you need to make a copy when marshaling them to Godot. There are methods in the Godot API that rely on them being mutable:

``` GDScript
var output = []
var pid = OS.execute('ls', [], true, output)
print(output.size()) # Output: 1
# output[0] contains the application standard output
```

``` C#
var output = new List<object>();
int pid = OS.Execute("ls", new string[] {}, true, output);
GD.Print(output.Count); // Output: 0
// Oops!
```

neikeq committed:
Add Dictionary::erase_checked(key) method
Same as erase, but it returns a boolean value indicating whether the pair was erased or not.
This method should be removed during the next compatibility breakage, and 'Dictionary::erase(key)' should be changed to return a boolean.

neikeq committed:
Add Array and Dictionary wrapper classes to C#

neikeq commented:
All right, this is done for now. I'll add PoolArray on a separate PR.
In order to avoid breaking compatibility, I added a new method, `erase_checked`, to Dictionary (not exposed to scripts) instead of modifying the return type of `erase`.

hpvb commented:
Cherry picked into 3.0.6
